enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,877 @@
+The validations will be same as mentioned in point C) except the one for passthrough routes.
+
+F) To validate the header value it has to match a regex.
+
+For ingress operator CR and route CR(microshift) this is defined on the request and response list of header in openshift/api.
+For route CR(OCP) this is defined in library-go.
+
+This will be the regex:
+```sh
+request header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+```sh
+response header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+
+Allowed dynamic values for request would be -
+```conf
+%[req.hdr(host),lower]
+%[req.hdr(host),base64]
+%[req.hdr(Host)] if foo
+```
+Allowed dynamic values for response would be -
+```conf
+%[res.hdr(X-Acl-Key),lower]
+%[res.hdr(X-Acl-Key),base64]
+%[req.hdr(Host)] if foo
+```
+Dynamic values allowed for both request and response would be -
+```conf
+%{+Q}[ssl_c_der,base64]
+```
+
+For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
+present in `pkg/route/validation/validation_test.go`
+
+Please note that all values will be sanitized before passing them to the haproxy.conf where a single quote present in value is escaped and the whole value is single quoted which will help
+haproxy not to crash in the event of an invalid input. This sanitize logic will be present in router.
+Quote using single quotes, we will do the following:
+Replace ' with '\''.
+Wrap the entire string in single quotes.
+eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''
+
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Would this enhancement cover setting or deleting request headers ?
+  Answer: Yes we would be covering setting or deleting of request headers.
+
+- Would we be able to set load balancing algorithm based on headers ?
+  Answer: No, this out of scope for this enhancement so, it will be considered as a non-goal. Please check point number 5 in the non-goals section for more details.
+
+- Would we be able to delete or append headers  ?
+  Answer: Yes, this enhancement will cover deletion and not appending of HTTP Response Headers as per the
+  [epic](https://issues.redhat.com/browse/RFE-464). 
+  The reason is "add-header" appends an HTTP header field whose name is specified in and whose value is defined by 
+  which follows the log-format rules. This may be used to send a cookie to a
+  client for example, or to pass some internal information. This rule is not final, so it is possible to add other 
+  similar rules. Note that header addition is performed immediately, so one rule might reuse the resulting header 
+  from a previous rule.
```

Comment:
You still haven't explained why appending isn't covered, so I'll make a guess. 
```suggestion
- Will this enhancement cover deleting or appending headers?
  Answer: Yes and no.  This enhancement will cover deletion but not appending of
  HTTP response headers as per the requirements from
  [RFE-464](https://issues.redhat.com/browse/RFE-464).  While this RFE mentions
  use-cases that require the ability to delete headers, there is no use-case
  from this RFE that requires the ability to append headers.  Once such a
  use-case arises, we may add the option to append headers, with due
  consideration of any additional requirements deriving from the use-case
  itself or considerations around the handling of comma-separated values.  (Note
  that a use-case involving the `x-forwarded-*` headers, which is not mentioned
  on the RFE, is described below.)
```

Reply:
This is mentioned in the EP -
  For the headers like `X-Forwarded-For` which might need values to be appended we have a different API for that.
  This API is specifically meant just either set or delete an HTTP Request and Response header.
  In this API if someone sets an HTTP header with a value we expect them to set a complete value in advance rather than appending it later to the header.